### PR TITLE
Disable 'Enable video' button if no camera is connected

### PR DESCRIPTION
**Pull request by bergart**
If theres no cam available, this fix will toggle the video icon into the disabled mode. This is solved by query the hardware from navigator.getUserMedia API. 

This Fix is based on:
Disable 'Enable video' button if no camera is connected [#441](https://github.com/nextcloud/spreed/issues/441)

selection_033

### CODE REVIEW

File: js/app.js [hunk 1]
```diff
@@ -175,7 +177,9 @@
 					// means that permission wasn't granted
 					// yet or there is no video available at
 					// all)
-					console.log('video can not be enabled - there was no stream available before');
+					console.log('video can not be enabled - there was no stream available before or no camera can be found');
+					// OCA.SpreedMe.app.availableVideo();
```
Comment (fancycode) [inline js/app.js]: Please remove any commented testing code.

File: js/app.js [hunk 2]
```diff
@@ -597,9 +601,34 @@
 				}
 			} else {
 				OCA.SpreedMe.app.videoWasEnabledAtLeastOnce = false;
-				OCA.SpreedMe.app.disableVideo();
+				// OCA.SpreedMe.app.availableAudio();
```
Comment (fancycode) [inline js/app.js]: Same here.

File: js/app.js [hunk 3]
```diff
@@ -597,9 +601,34 @@
...
+				OCA.SpreedMe.app.availableVideo();
 			}
 		},
+		//// INFO: Not needed: no microphone results in an access denied failure
```
Comment (fancycode) [inline js/app.js]: And here. Also what happens if a user has video but no audio?

File: js/app.js [hunk 4]
```diff
@@ -50,6 +50,8 @@
 		displayedGuestNameHint: false,
 		audioDisabled: localStorage.getItem("audioDisabled"),
 		videoDisabled: localStorage.getItem("videoDisabled"),
+		audioAvailable: localStorage.getItem("audioAvailable"),
```
Comment (fancycode) [inline js/app.js]: You are not saving this anywhere.

File: js/app.js [hunk 5]
```diff
@@ -616,6 +645,30 @@
...
 		},
+		availableVideo: function() {
+			var $hideVideoButton = $('#hideVideo');
+
+			navigator.getUserMedia = navigator.getUserMedia ||
+                         navigator.webkitGetUserMedia ||
+                         navigator.mozGetUserMedia;
+
+			if (navigator.getUserMedia !== null) {
+				navigator.getUserMedia({ video: true },
```
Comment (fancycode) [inline js/app.js]: There should be no need to call `getUserMedia` again, this was already done inside `SimpleWebRTC`. Also the streams here could be different from what was returned to the original `getUserMedia`.

File: js/app.js [hunk 6]
```diff
@@ -616,6 +645,30 @@
...
+					function() {
+						OCA.SpreedMe.app.enableVideo();
+						localStorage.removeItem("videoDisabled");
+						OCA.SpreedMe.app.videoAvailable = true;
+					},
+		      function(err) {
+						$hideVideoButton.attr('data-original-title', 'No Camera')
```
Comment (fancycode) [inline js/app.js]: Please make this translateable.

**Comment by fancycode** [comment]
Also please check the JS syntax errors as reported by eslint.

File: js/app.js (hunk 3 above)
Comment (bergart) [reply js/app.js]: OK, the comments will be removed. If there is Video but no Audio (microphone) the session wont be initialized! This is a bug, a issue is opened!

File: js/app.js (hunk 5 above)
Comment (bergart) [reply js/app.js]: you are right!

File: js/app.js (hunk 6 above)
Comment (bergart) [reply js/app.js]: ohh i see it, i will also change the other ones.

File: js/app.js [hunk 7]
```diff
@@ -211,7 +214,7 @@
 					} else if (fullscreenElem.msRequestFullscreen) {
 						fullscreenElem.msRequestFullscreen();
 					}
-					$(this).attr('data-original-title', 'Exit fullscreen');
+					$(this).attr('data-original-title', t('spreed', 'Exit fullscreen'));
```
Comment (fancycode) [inline js/app.js]: Please don't change unrelated parts of the file, this should be in a separate PR (applies to other parts below, too).

File: js/app.js [hunk 8]
```diff
@@ -597,32 +600,43 @@
...
+			$('#mute').attr('data-original-title', t('spreed', 'Enable audio'))
 				.addClass('audio-disabled icon-audio-off-white')
 				.removeClass('icon-audio-white');
 
 			OCA.SpreedMe.app.audioDisabled = true;
 		},
+		availableVideo: function() {
+			var $hideVideoButton = $('#hideVideo');
+
+			if (OCA.SpreedMe.webrtc.localStreams !== null) {
```
Comment (fancycode) [inline js/app.js]: Did you actually test this? `OCA.SpreedMe.webrtc.localStreams` should be a list in which the last entry is the new local stream. You then need to check if this stream contains an audio and/or video track.

**Comment by bergart** [comment]
Fix was revised.

File: js/webrtc.js [hunk 9]
```diff
@@ -931,6 +931,14 @@ var spreedPeerConnectionTable = [];
 			if (!OCA.SpreedMe.app.videoDisabled) {
 				OCA.SpreedMe.app.enableVideo();
 			}
+
+			var availableStream = $(OCA.SpreedMe.webrtc.webrtc.localStream.getVideoTracks());
```
Comment (fancycode) [inline js/webrtc.js]: No need to wrap in a jQuery object.

File: js/webrtc.js [hunk 10]
```diff
@@ -931,6 +931,14 @@ var spreedPeerConnectionTable = [];
...
+			var $hideVideoButton = $('#hideVideo');
+			if(availableStream.length < 1 || availableStream.length == undefined) {
```
Comment (fancycode) [inline js/webrtc.js]: You could write this simpler as `if (!availableStream.length) {` (btw, this is an array, so `length` is never undefined).

File: css/style.css [hunk 11]
```diff
@@ -432,6 +432,15 @@ video {
 	opacity: .7;
 }
 
+.nameIndicator button.video-available {
```
Comment (fancycode) [inline css/style.css]: The class is added, if no video track is in the stream, right? If so, the name of the class should not be `video-available` but something like `no-video-tracks` / `no-video-available` (or similar).

Comment (bergart) [reply css/style.css]: no-video-tracks sounds good to me

File: js/webrtc.js [hunk 12]
```diff
@@ -931,6 +931,13 @@ var spreedPeerConnectionTable = [];
 			if (!OCA.SpreedMe.app.videoDisabled) {
 				OCA.SpreedMe.app.enableVideo();
 			}
+
+			var $hideVideoButton = $('#hideVideo');
+			if(OCA.SpreedMe.webrtc.webrtc.localStream.getVideoTracks() < 1) {
```
Comment (fancycode) [inline js/webrtc.js]: Shouldn't that be `OCA.SpreedMe.webrtc.webrtc.localStream.getVideoTracks().length < 1`?

File: js/webrtc.js [hunk 13]
```diff
@@ -931,6 +931,13 @@ var spreedPeerConnectionTable = [];
...
+				$hideVideoButton.removeClass('video-disabled icon-video-off-white')
+				.addClass('no-video-available icon-video-off-white')
```
Comment (fancycode) [inline js/webrtc.js]: You remove `icon-video-off-white` and add it immediately back, this is unnecessary.

**Comment by fancycode** [comment]
Another thing to test to check if everything is reset back properly:
1. Join call with camera disabled -> UI should disable the video button
2. Switch to another call with camera enabled (this doesn't reload the app) -> UI should enable the video button

**Comment by bergart** [comment]
@fancycode 

> Join call with camera disabled -> UI should disable the video button
tested and it works

> Switch to another call with camera enabled (this doesn't reload the app) -> UI should enable the video button
also tested and it works

**Comment by nickvergessen** [comment]
Couldn't really test with my laptop because of the build in camera, but I made sure `OCA.SpreedMe.webrtc.webrtc.localStream.getVideoTracks().length` returns 1 for me and tested it with that and it correctly disabled the button.

Thanks for your contribution.
